pkg/controllers/nodeipam/ipam/cidr_allocator.go

```diff
@@ -0,0 +1,480 @@
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.nodesInProcessing.Delete(nodeName)
+}
+
+// marks node.PodCIDRs[...] as used in allocator's tracked cidrSet
+func (r *rangeAllocator) occupyCIDRs(node *v1.Node) error {
+	defer r.removeNodeFromProcessing(node.Name)
+	if len(node.Spec.PodCIDRs) == 0 {
+		return nil
+	}
+	for idx, cidr := range node.Spec.PodCIDRs {
+		_, podCIDR, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return fmt.Errorf("failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
+		}
+		// If node has a pre allocate cidr that does not exist in our cidrs.
+		// This will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
+			return fmt.Errorf("node:%s has an allocated cidr: %v at index:%v that does not exist in cluster cidrs configuration", node.Name, cidr, idx)
+		}
+
+		if err := r.cidrSets[idx].Occupy(podCIDR); err != nil {
+			return fmt.Errorf("failed to mark cidr[%v] at idx [%v] as occupied for node: %v: %v", podCIDR, idx, node.Name, err)
+		}
+	}
+	return nil
+}
+
+// WARNING: If you're adding any return calls or defer any more work from this
+// function you have to make sure to update nodesInProcessing properly with the
+// disposition of the node when the work is done.
+func (r *rangeAllocator) AllocateOrOccupyCIDR(node *v1.Node) error {
+	if node == nil {
+		return nil
+	}
+	if !r.insertNodeToProcessing(node.Name) {
+		klog.V(2).Infof("Node %v is already in a process of CIDR assignment.", node.Name)
+		return nil
+	}
+
+	if len(node.Spec.PodCIDRs) > 0 {
+		return r.occupyCIDRs(node)
+	}
+	// allocate and queue the assignment
+	allocated := NodeReservedCIDRs{
+		nodeName:       node.Name,
+		allocatedCIDRs: make([]*net.IPNet, len(r.cidrSets)),
+	}
+
+	for idx := range r.cidrSets {
+		podCIDR, err := r.cidrSets[idx].AllocateNext()
+		if err != nil {
+			r.removeNodeFromProcessing(node.Name)
+			nodeutil.RecordNodeStatusChange(r.recorder, node, "CIDRNotAvailable")
+			return fmt.Errorf("failed to allocate cidr from cluster cidr at idx:%v: %v", idx, err)
+		}
+		allocated.allocatedCIDRs[idx] = podCIDR
+	}
+
+	//queue the assignment
+	klog.V(4).Infof("Putting node %s with CIDR %v into the work queue", node.Name, allocated.allocatedCIDRs)
+	r.nodeCIDRUpdateChannel <- allocated
+	return nil
+}
+
+// ReleaseCIDR marks node.podCIDRs[...] as unused in our tracked cidrSets
+func (r *rangeAllocator) ReleaseCIDR(node *v1.Node) error {
+	if node == nil || len(node.Spec.PodCIDRs) == 0 {
+		return nil
+	}
+
+	for idx, cidr := range node.Spec.PodCIDRs {
+		_, podCIDR, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return fmt.Errorf("failed to parse CIDR %s on Node %v: %v", cidr, node.Name, err)
+		}
+
+		// If node has a pre allocate cidr that does not exist in our cidrs.
+		// This will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
```

Comment:
same as previous comment, i'd like to move this check before the loop for performance and readability.
the len(node.spec.PodCIDRs)(when it's not 0) must always **equal to**(instead of `<=`)  len(r.cidrSets)